crystal_diffusion/data/diffusion/data_loader.py

```diff
@@ -0,0 +1,167 @@
+        transformed_x['natom'] = torch.as_tensor(x['natom']).long()  # resulting tensor size: (batchsize, )
+        bsize = transformed_x['natom'].size(0)
+        transformed_x['box'] = torch.as_tensor(x['box'])  # size: (batchsize, spatial dimension)
+        for pos in ['position', 'reduced_position']:
+            transformed_x[pos] = torch.as_tensor(x[pos]).view(bsize, -1, spatial_dim)
+        transformed_x['type'] = torch.as_tensor(x['type']).long()  # size: (batchsize, natom after padding)
+
+        return transformed_x
+
+    @staticmethod
+    def pad_samples(x: Dict[typing.AnyStr, typing.Any], max_atom: int, spatial_dim: int = 3) -> Dict[str, torch.Tensor]:
+        """Pad a sample for batching.
+
+        Args:
+            x: initial sample from the dataset. Should contain natom, position, and type.
+            max_atom: maximum number of atoms to pad to
+            spatial_dim (optional): number of spatial dimensions. Defaults to 3.
+
+        Returns:
+            x: sample with padded type and position
+        """
+        natom = x['natom']
+        if natom > max_atom:
+            raise ValueError(f"Hyper-parameter max_atom is smaller than an example in the dataset with {natom} atoms.")
+        x['type'] = F.pad(torch.as_tensor(x['type']).long(), (0, max_atom - natom), 'constant', -1)
+        for pos in ['position', 'reduced_position']:
+            x[pos] = F.pad(torch.as_tensor(x[pos]).float(), (0, spatial_dim * (max_atom - natom)), 'constant',
+                           torch.nan)
+        return x
+
+    def setup(self, stage: Optional[str] = None):
+        """Parse and split all samples across the train/valid/test parsers."""
+        # here, we will actually assign train/val datasets for use in dataloaders
+        processed_data = LammpsProcessorForDiffusion(self.lammps_run_dir, self.processed_dataset_dir)
+
+        if stage == "fit" or stage is None:
+            self.train_dataset = datasets.Dataset.from_parquet(processed_data.train_files,
+                                                               cache_dir=self.working_cache_dir)
+            self.valid_dataset = datasets.Dataset.from_parquet(processed_data.valid_files,
+                                                               cache_dir=self.working_cache_dir)
+            # TODO QoL valid dataset is labelled as train split by Datasets. Find a way to rename.
+        else:
+            raise NotImplementedError("Test mode needs to be implemented.")
+        # TODO test dataset when stage == 'test'
+        # we can filter out samples at this stage using the .filter(lambda x: f(x)) with f(x) a boolean function
+        # or a .select(list[int]) with a list of indices to keep a subset. This is much faster than .filter
+        # padding needs to be done here OR in the preprocessor
+        # check if the max number of atoms matches at least the max in the training set
+        self.max_atom = max(max(self.train_dataset['natom']), self.max_atom)
```

Comment:
Here you are potentially changing an input parameter... I would prefer an assert that forces the user to set a correct value. Alternatively, send a warning through the `Logger`?